pkg/cache/application.go

```diff
@@ -363,6 +406,52 @@ func (app *Application) handleRecoverApplicationEvent(event *fsm.Event) {
 	}
 }
 
+func (app *Application) postAppAccepted(event *fsm.Event) {
+	// if app has taskGroups defined, it goes to the Reserving state before getting to Running
+	var ev events.SchedulingEvent
+	if len(app.taskGroups) != 0 {
+		ev = NewSimpleApplicationEvent(app.applicationID, events.TryReserve)
+		dispatcher.Dispatch(ev)
+	} else {
+		ev = NewRunApplicationEvent(app.applicationID)
+	}
+	dispatcher.Dispatch(ev)
+}
+
+func (app *Application) onReserving(event *fsm.Event) {
+	go func() {
+		// while doing reserving
+		if err := GetPlaceholderManager().createAppPlaceholders(app); err != nil {
+			// creating placeholder failed
+			// put the app into recycling queue and turn the app to running state
+			GetPlaceholderManager().CleanUp(app)
+			ev := NewRunApplicationEvent(app.applicationID)
```

Comment:
Correct. This is just in case we get some failures while creating placeholders. In most cases, this should not happen.
When this happens, today, we do not have a good understanding of what should be the best way to handle such failures. So we take a simple and safe solution, aka release all the placeholders and try to schedule the app as a normal app. We can think about how to improve this when we know what is the best optimization.